File: ios/Plugin/DBRPlugin.swift
import Foundation
import Capacitor
import DynamsoftBarcodeReader
import DynamsoftCameraEnhancer

/**
 * Please read the Capacitor iOS Plugin Development Guide
 * here: https://capacitorjs.com/docs/plugins/ios
 */
@objc(DBRPlugin)
public class DBRPlugin: CAPPlugin, DBRLicenseVerificationListener, DCELicenseVerificationListener  {

    private let implementation = DBR()
    var dce:DynamsoftCameraEnhancer! = nil
    var dceView:DCECameraView! = nil
    var barcodeReader:DynamsoftBarcodeReader! = nil
    var timer:Timer! = nil;
    
    @objc func destroy(_ call: CAPPluginCall) {
        if (barcodeReader == nil) {
            call.reject("not initialized")
        }else{
            barcodeReader = nil
            DispatchQueue.main.sync {
                dceView.removeFromSuperview()
            }
            dceView = nil
            dce = nil
            call.resolve()
        }
    }
    
    @objc func initialize(_ call: CAPPluginCall) {
        if (barcodeReader == nil){
            let license = call.getString("license") ?? "DLS2eyJvcmdhbml6YXRpb25JRCI6IjIwMDAwMSJ9"
            let dceLicense = call.getString("dceLicense") ?? "DLS2eyJvcmdhbml6YXRpb25JRCI6IjIwMDAwMSJ9"
            
            configurationDBR(license: license)
            configurationDCE(license: dceLicense)
        }
        var ret = PluginCallResultData()
        ret["success"] = true
        call.resolve(ret)
    }
    
    @objc func initRuntimeSettingsWithString(_ call: CAPPluginCall) {
        let template = call.getString("template") ?? ""
        if (template != ""){
            try? barcodeReader.initRuntimeSettingsWithString(template, conflictMode: EnumConflictMode.overwrite)
        }
        call.resolve()
    }
    
    @objc func startScan(_ call: CAPPluginCall) {
        makeWebViewTransparent()
        if dce != nil {
            DispatchQueue.main.sync {
                dce.open()
                triggerOnPlayed()
                if timer != nil {
                    timer.invalidate()
                }
                timer = Timer.scheduledTimer(timeInterval: 0.1, target: self, selector: #selector(decodingTask), userInfo: nil, repeats: true)
            }
        }else{
            call.reject("not initialized")
            return
        }
        call.resolve()
    }
    
    func makeWebViewTransparent(){
        DispatchQueue.main.async {
           self.bridge?.webView!.isOpaque = false
           self.bridge?.webView!.backgroundColor = UIColor.clear
           self.bridge?.webView!.scrollView.backgroundColor = UIColor.clear
       }
    }
    func restoreWebViewBackground(){
        DispatchQueue.main.async {
           self.bridge?.webView!.isOpaque = true
           self.bridge?.webView!.backgroundColor = UIColor.white
           self.bridge?.webView!.scrollView.backgroundColor = UIColor.white
       }
    }
    
    @objc func toggleTorch(_ call: CAPPluginCall) {
        if (dce == nil){
            call.reject("not initialized")
        }else{
            if call.getBool("on", true){
               dce.turnOnTorch()
            } else{
               dce.turnOffTorch()
            }
            call.resolve()
        }
    }
    
    @objc func stopScan(_ call: CAPPluginCall) {
        restoreWebViewBackground()
        if (dce == nil){
            call.reject("not initialized")
        }else{
            dce.close()
            call.resolve()
        }
    }
    
    @objc func resumeScan(_ call: CAPPluginCall) {
        if (dce == nil){
            call.reject("not initialized")
        }else{
            dce.resume()
            call.resolve()
        }
    }
    
    @objc func pauseScan(_ call: CAPPluginCall) {
        if (dce == nil){
            call.reject("not initialized")
        }else{
            dce.pause()
            call.resolve()
        }
    }
    
    @objc func setResolution(_ call: CAPPluginCall) {
        if (dce == nil){
            call.reject("not initialized")
        }else{
            let res = call.getInt("resolution") ?? -1
            NSLog("Resolution: %d", res)
            
            if res != -1 {
                let resolution = EnumResolution.init(rawValue: res)
                dce.setResolution(resolution!)
                triggerOnPlayed()
            }
            
            var ret = PluginCallResultData()
            ret["success"] = true
            call.resolve(ret)
        }
    }
    
    @objc func getResolution(_ call: CAPPluginCall) {
        if (dce == nil){
            call.reject("not initialized")
        }else{
            var ret = PluginCallResultData()
            let res = dce.getResolution();
            dce.getResolution()
            print("res: "+res)
            ret["resolution"] = res
            call.resolve(ret)
        }
    }
    
    @objc func triggerOnPlayed() {
        if (dce != nil) {
            var ret = PluginCallResultData()
            let res = dce.getResolution()
            ret["resolution"] = res
            print("trigger on played")
            notifyListeners("onPlayed", data: ret)
        }
    }
    
    @objc func getAllCameras(_ call: CAPPluginCall) {
        if (dce == nil){
            call.reject("not initialized")
        }else{
            var ret = PluginCallResultData()
            let array = NSMutableArray();
            array.addObjects(from: dce.getAllCameras())
            ret["cameras"] = array
            call.resolve(ret)
        }
    }
    
    @objc func getSelectedCamera(_ call: CAPPluginCall) {
        if (dce == nil){
            call.reject("not initialized")
        }else{
            var ret = PluginCallResultData()
            ret["selectedCamera"] = dce.getSelectedCamera()
            call.resolve(ret)
        }
    }
    
    @objc func selectCamera(_ call: CAPPluginCall) {
        if (dce == nil){
            call.reject("not initialized")
        }else{
            let cameraID = call.getString("cameraID") ?? ""
            if cameraID != "" {
                var error: NSError? = NSError()
                dce.selectCamera(cameraID, error: &error)
                triggerOnPlayed()
            }
            var ret = PluginCallResultData()
            ret["success"] = true
            call.resolve(ret)
        }
    }
    
    @objc func setScanRegion(_ call: CAPPluginCall) {
        if (dce == nil){
            call.reject("not initialized")
        }else{
            let scanRegion = iRegionDefinition()
            scanRegion.regionTop = call.getInt("top")!
            scanRegion.regionBottom = call.getInt("bottom")!
            scanRegion.regionLeft = call.getInt("left")!
            scanRegion.regionRight = call.getInt("right")!
            scanRegion.regionMeasuredByPercentage = call.getInt("measuredByPercentage")!
            var error: NSError? = NSError()
            dce.setScanRegion(scanRegion, error: &error)
            var ret = PluginCallResultData()
            ret["success"] = true
            call.resolve(ret)
        }
    }
    
    @objc func setZoom(_ call: CAPPluginCall) {
        if (dce == nil){
            call.reject("not initialized")
        }else{
            let factor:CGFloat = CGFloat(call.getFloat("factor") ?? 1.0)
            dce.setZoom(factor)
            var ret = PluginCallResultData()
            ret["success"] = true
            call.resolve(ret)
        }
    }
    
    @objc func setFocus(_ call: CAPPluginCall) {
        if (dce == nil){
            call.reject("not initialized")
        }else{
            let x = call.getFloat("x", -1.0);
            let y = call.getFloat("y", -1.0);
            if x != -1.0 && y != -1.0 {
                dce.setFocus(CGPoint(x: CGFloat(x), y: CGFloat(y)))
            }
            var ret = PluginCallResultData()
            ret["success"] = true
            call.resolve(ret)
        }
    }
    
    func configurationDBR(license: String) {
        DynamsoftBarcodeReader.initLicense(license, verificationDelegate: self)
        barcodeReader = DynamsoftBarcodeReader.init()
    }
    
    public func dbrLicenseVerificationCallback(_ isSuccess: Bool, error: Error?)
    {
        let err = error as NSError?
        if(err != nil){
            print("Server DBR license verify failed")
        }
    }
    
    func configurationDCE(license: String) {
        DynamsoftCameraEnhancer.initLicense(license,verificationDelegate:self)

        // Initialize a camera view for previewing video.
        DispatchQueue.main.sync {
            print("configuring dce")
            dceView = DCECameraView.init(frame: (bridge?.viewController?.view.bounds)!)
            dceView.overlayVisible = true
            self.webView!.superview!.insertSubview(dceView, belowSubview: self.webView!)
            dce = DynamsoftCameraEnhancer.init(view: dceView)
            dce.setResolution(EnumResolution.EnumRESOLUTION_720P)
        }
    }
    
    public func dceLicenseVerificationCallback(_ isSuccess: Bool, error: Error?) {
        let err = error as NSError?
        if(err != nil){
            print("Server DCE license verify failed")
        }
    }

    @objc func decodingTask()
    {
        if dce == nil {
            return
        }
        
        if dce.getCameraState() != EnumCameraState.EnumCAMERA_STATE_OPENED {
            return
        }
        
        let frame = dce.getFrameFromBuffer(false)
        let results = try? barcodeReader.decodeBuffer(frame.imageData, width: frame.width, height: frame.height, stride: frame.stride, format: EnumImagePixelFormat(rawValue: frame.pixelFormat) ?? EnumImagePixelFormat.NV21)
        let count = results?.count ?? 0
        NSLog("Found %d barcode(s)", count)
        var ret = PluginCallResultData()
        let array = NSMutableArray();
        for index in 0..<count {
            let tr = results![index]
            let points = tr.localizationResult?.resultPoints as! [CGPoint]
            var result = PluginCallResultData()
            result["barcodeText"] = tr.barcodeText
            result["barcodeFormat"] = tr.barcodeFormatString
            result["barcodeBytesBase64"] = tr.barcodeBytes?.base64EncodedString()
            for j in 0..<4 {
                var x = points[j].x
                var y = points[j].y
                if frame.isCropped {
                    x = frame.cropRegion.minX + x
                    y = frame.cropRegion.minY + y
                }
                result["x"+String(j+1)] = x
                result["y"+String(j+1)] = y
            }
            array.add(result)
        }
        ret["results"]=array
        ret["frameOrientation"] = frame.orientation
        if UIApplication.shared.statusBarOrientation.isLandscape {
            ret["deviceOrientation"] = "landscape"
        }else {
            ret["deviceOrientation"] = "portrait"
        }
        notifyListeners("onFrameRead", data: ret)
    }
}

File: android/src/main/java/com/dynamsoft/capacitor/DBRPlugin.java
package com.dynamsoft.capacitor;

import android.content.res.Configuration;
import android.graphics.Color;
import android.graphics.drawable.Drawable;
import android.util.Base64;
import android.util.Log;
import android.util.Size;
import android.view.View;
import android.view.ViewGroup;
import android.widget.FrameLayout;

import com.dynamsoft.dbr.BarcodeReader;
import com.dynamsoft.dbr.BarcodeReaderException;
import com.dynamsoft.dbr.DBRLicenseVerificationListener;
import com.dynamsoft.dbr.EnumConflictMode;
import com.dynamsoft.dbr.TextResult;
import com.dynamsoft.dce.CameraEnhancer;
import com.dynamsoft.dce.CameraEnhancerException;
import com.dynamsoft.dce.DCECameraView;
import com.dynamsoft.dce.DCEFrame;
import com.dynamsoft.dce.DCELicenseVerificationListener;
import com.dynamsoft.dce.EnumCameraState;
import com.dynamsoft.dce.EnumResolution;
import com.getcapacitor.JSArray;
import com.getcapacitor.JSObject;
import com.getcapacitor.Plugin;
import com.getcapacitor.PluginCall;
import com.getcapacitor.PluginMethod;
import com.getcapacitor.annotation.CapacitorPlugin;

import java.util.Timer;
import java.util.TimerTask;

@CapacitorPlugin(name = "DBR")
public class DBRPlugin extends Plugin {

    private DBR implementation = new DBR();
    private CameraEnhancer mCameraEnhancer = null;
    private DCECameraView mCameraView;
    private BarcodeReader reader = null;
    private Timer timer = null;
    private EnumCameraState previousCameraStatus = null;

    private void startDecodingTimer(){
        timer = new Timer();
        TimerTask task = new TimerTask() {
            @Override
            public void run() {
                if (mCameraEnhancer!=null) {
                    if (mCameraEnhancer.getCameraState() == EnumCameraState.OPENED){
                        DCEFrame frame = mCameraEnhancer.getFrameFromBuffer(false);
                        if (frame != null){
                            try {
                                TextResult[] textResults = reader.decodeBuffer(frame.getImageData(),frame.getWidth(),frame.getHeight(),frame.getStrides()[0],frame.getPixelFormat());
                                JSObject ret = wrapResults(textResults, frame);
                                ret.put("frameOrientation",frame.getOrientation());
                                int deviceOrientation = getContext().getResources().getConfiguration().orientation;
                                if (deviceOrientation == Configuration.ORIENTATION_PORTRAIT) {
                                    ret.put("deviceOrientation", "portrait");
                                }else{
                                    ret.put("deviceOrientation", "landscape");
                                }
                                Log.d("DBR","Found "+textResults.length+" barcode(s).");
                                notifyListeners("onFrameRead",ret);
                            } catch (BarcodeReaderException e) {
                                e.printStackTrace();
                            }
                        }
                    }
                }
            }
        };
        timer.scheduleAtFixedRate(task, 1000, 100);
    }

    @PluginMethod
    public void destroy(PluginCall call) {
        if (reader!=null){
            getActivity().runOnUiThread(new Runnable() {
                public void run() {
                    Log.d("DBR","destroy");
                    ((ViewGroup) bridge.getWebView().getParent()).removeView(mCameraView);
                    mCameraView = null;
                    mCameraEnhancer = null;
                    reader = null;
                }
            });
        }
        call.resolve();
    }

    @PluginMethod
    public void initialize(PluginCall call) {
        try {
            if (reader==null) {
                initDBR(call);
            }
            Runnable dceInit = new Runnable() {
                public void run() {
                    if (mCameraEnhancer == null) {
                        String dceLicense = call.getString("dceLicense","DLS2eyJvcmdhbml6YXRpb25JRCI6IjIwMDAwMSJ9");
                        initDCE(dceLicense);
                    }
                    synchronized(this)
                    {
                        this.notify();
                    }
                }
            };
            synchronized( dceInit ) {
                getActivity().runOnUiThread(dceInit);
                dceInit.wait() ; // unlocks myRunable while waiting
            }
        } catch (BarcodeReaderException | InterruptedException e) {
            e.printStackTrace();
            call.reject(e.getMessage());
        }
        JSObject result = new JSObject();
        result.put("success",true);
        call.resolve(result);
    }

    @PluginMethod
    public void initRuntimeSettingsWithString(PluginCall call){
        if (call.hasOption("template")){
            try {
                reader.initRuntimeSettingsWithString(call.getString("template"), EnumConflictMode.CM_OVERWRITE);
            } catch (BarcodeReaderException e) {
                e.printStackTrace();
                call.reject(e.getMessage());
                return;
            }
        }
        call.resolve();
    }

    @PluginMethod
    public void selectCamera(PluginCall call){
        if (call.hasOption("cameraID")){
            try {
                Runnable selectCameraRunnable = new Runnable() {
                    public void run() {
                        try {
                            mCameraEnhancer.selectCamera(call.getString("cameraID"));
                        } catch (CameraEnhancerException e) {
                            e.printStackTrace();
                        }
                        synchronized(this)
                        {
                            this.notify();
                        }
                    }
                };
                synchronized( selectCameraRunnable ) {
                    getActivity().runOnUiThread(selectCameraRunnable);
                    selectCameraRunnable.wait();
                }

                triggerOnPlayed();
            } catch (InterruptedException e) {
                e.printStackTrace();
                call.reject(e.getMessage());
            }
        }
        JSObject result = new JSObject();
        result.put("success",true);
        call.resolve(result);
    }

    private void triggerOnPlayed(){
        Size res = mCameraEnhancer.getResolution();
        if (res != null) {
            JSObject onPlayedResult = new JSObject();
            onPlayedResult.put("resolution",res.getWidth() + "x" + res.getHeight());
            Log.d("DBR","resolution:" + res.getWidth() + "x" + res.getHeight());
            notifyListeners("onPlayed",onPlayedResult);
        }
    }

    @PluginMethod
    public void getAllCameras(PluginCall call){
        if (mCameraEnhancer == null) {
            call.reject("not initialized");
        }else {
            JSObject result = new JSObject();
            JSArray cameras = new JSArray();
            for (String camera: mCameraEnhancer.getAllCameras()) {
                cameras.put(camera);
            }
            result.put("cameras",cameras);
            call.resolve(result);
        }
    }

    @PluginMethod
    public void getSelectedCamera(PluginCall call){
        if (mCameraEnhancer == null) {
            call.reject("not initialized");
        }else{
            JSObject result = new JSObject();
            result.put("selectedCamera",mCameraEnhancer.getSelectedCamera());
            call.resolve(result);
        }
    }

    @PluginMethod
    public void setResolution(PluginCall call){
        if (call.hasOption("resolution")){
            try {
                Runnable setResolutionRunnable = new Runnable() {
                    public void run() {
                        try {
                            mCameraEnhancer.setResolution(EnumResolution.fromValue(call.getInt("resolution")));
                        } catch (CameraEnhancerException e) {
                            e.printStackTrace();
                        }
                        synchronized(this)
                        {
                            this.notify();
                        }
                    }
                };
                synchronized( setResolutionRunnable ) {
                    getActivity().runOnUiThread(setResolutionRunnable);
                    setResolutionRunnable.wait();
                }
                triggerOnPlayed();
            } catch (InterruptedException e) {
                e.printStackTrace();
                call.reject(e.getMessage());
            }
        }
        JSObject result = new JSObject();
        result.put("success",true);
        call.resolve(result);
    }

    @PluginMethod
    public void getResolution(PluginCall call){
        if (mCameraEnhancer == null) {
            call.reject("not initialized");
        }else{
            String res = mCameraEnhancer.getResolution().getWidth()+"x"+mCameraEnhancer.getResolution().getHeight();
            JSObject result = new JSObject();
            result.put("resolution",res);
            call.resolve(result);
        }
    }

    @PluginMethod
    public void startScan(PluginCall call) {
        getActivity().runOnUiThread(new Runnable() {
            public void run() {
                try {
                    mCameraView.setVisibility(View.VISIBLE);
                    mCameraEnhancer.open();
                    makeWebViewTransparent();
                    triggerOnPlayed();
                    startDecodingTimer();
                    call.resolve();
                } catch (CameraEnhancerException e) {
                    e.printStackTrace();
                    call.reject(e.getMessage());
                }
            }
        });

    }

    @PluginMethod
    public void setScanRegion(PluginCall call){
        if (mCameraEnhancer!=null) {
            getActivity().runOnUiThread(new Runnable() {
                public void run() {
                    com.dynamsoft.dce.RegionDefinition scanRegion = new com.dynamsoft.dce.RegionDefinition();
                    scanRegion.regionTop = call.getInt("top");
                    scanRegion.regionBottom = call.getInt("bottom");
                    scanRegion.regionLeft = call.getInt("left");
                    scanRegion.regionRight = call.getInt("right");
                    scanRegion.regionMeasuredByPercentage = call.getInt("measuredByPercentage");

                    try {
                        mCameraEnhancer.setScanRegion(scanRegion);
                        call.resolve();
                    } catch (CameraEnhancerException e) {
                        e.printStackTrace();
                        call.reject(e.getMessage());
                    }
                }
            });
        }else{
            call.reject("not initialized");
        }
    }

    @PluginMethod
    public void setZoom(PluginCall call){
        if (call.hasOption("factor")) {
            Float factor = call.getFloat("factor");
            try {
                mCameraEnhancer.setZoom(factor);
            } catch (CameraEnhancerException e) {
                e.printStackTrace();
                call.reject(e.getMessage());
            }
        }
        call.resolve();
    }

    @PluginMethod
    public void setFocus(PluginCall call){
        if (call.hasOption("x") && call.hasOption("y")) {
            Float x = call.getFloat("x");
            Float y = call.getFloat("y");
            try {
                mCameraEnhancer.setFocus(x,y);
            } catch (CameraEnhancerException e) {
                e.printStackTrace();
                call.reject(e.getMessage());
            }
        }
        call.resolve();
    }


    private void initDBR(PluginCall call) throws BarcodeReaderException {
        BarcodeReader.initLicense(call.getString("license","DLS2eyJvcmdhbml6YXRpb25JRCI6IjIwMDAwMSJ9"), new DBRLicenseVerificationListener() {
            @Override
            public void DBRLicenseVerificationCallback(boolean isSuccessful, Exception e) {
                if (!isSuccessful) {
                    e.printStackTrace();
                }
            }
        });
        reader = new BarcodeReader();
    }

    private void initDCE(String dceLicense){
        CameraEnhancer.initLicense(dceLicense, new DCELicenseVerificationListener() {
            @Override
            public void DCELicenseVerificationCallback(boolean isSuccess, Exception error) {
                if(!isSuccess){
                    error.printStackTrace();
                }
            }
        });

        mCameraEnhancer = new CameraEnhancer(getActivity());
        mCameraView = new DCECameraView(getActivity());
        mCameraEnhancer.setCameraView(mCameraView);

        FrameLayout.LayoutParams cameraPreviewParams = new FrameLayout.LayoutParams(
                FrameLayout.LayoutParams.WRAP_CONTENT,
                FrameLayout.LayoutParams.WRAP_CONTENT
        );
        ((ViewGroup) bridge.getWebView().getParent()).addView(mCameraView,cameraPreviewParams);
        bridge.getWebView().bringToFront();
    }

    private JSObject wrapResults(TextResult[] results, DCEFrame frame) {
        JSObject ret = new JSObject();
        JSArray array = new JSArray();
        for (TextResult result:results){
            JSObject oneRet = new JSObject();
            oneRet.put("barcodeText", result.barcodeText);
            oneRet.put("barcodeFormat", result.barcodeFormatString);
            oneRet.put("barcodeBytesBase64", Base64.encodeToString(result.barcodeBytes,Base64.DEFAULT));

            for (int i = 0; i < 4 ; i++) {
                int x = result.localizationResult.resultPoints[i].x;
                int y = result.localizationResult.resultPoints[i].y;
                if (frame.getIsCropped()) {
                    x = x + frame.getCropRegion().left;
                    y = y + frame.getCropRegion().top;
                }
                oneRet.put("x"+(i+1), x);
                oneRet.put("y"+(i+1), y);
            }
            array.put(oneRet);
        }
        ret.put("results",array);
        return ret;
    }

    private void makeWebViewTransparent(){
        bridge.getWebView().setTag(bridge.getWebView().getBackground());
        bridge.getWebView().setBackgroundColor(Color.TRANSPARENT);
        //bridge.getWebView().evaluateJavascript("document.body.style.display='none'",null);
    }

    private void restoreWebViewBackground(){
        bridge.getWebView().setBackground((Drawable) bridge.getWebView().getTag());
    }

    @PluginMethod
    public void toggleTorch(PluginCall call) {
        try{
            if (call.getBoolean("on",true)){
                mCameraEnhancer.turnOnTorch();
            }else {
                mCameraEnhancer.turnOffTorch();
            }
            call.resolve();
        }catch (Exception e){
            call.reject(e.getMessage());
        }
    }

    @PluginMethod
    public void stopScan(PluginCall call) {
        try{
            restoreWebViewBackground();
            mCameraView.setVisibility(View.INVISIBLE);
            mCameraEnhancer.close();
            if (timer != null) {
                timer.cancel();
            }
            call.resolve();
        }catch (Exception e){
            call.reject(e.getMessage());
        }
    }

    @PluginMethod
    public void pauseScan(PluginCall call) {
        try{
            mCameraEnhancer.pause();
            call.resolve();
        }catch (Exception e){
            call.reject(e.getMessage());
        }
    }

    @PluginMethod
    public void resumeScan(PluginCall call) {
        try{
            mCameraEnhancer.resume();
            call.resolve();
        }catch (Exception e){
            call.reject(e.getMessage());
        }
    }


    @Override
    protected void handleOnPause() {
        if (mCameraEnhancer!=null){
            try {
                previousCameraStatus = mCameraEnhancer.getCameraState();
                mCameraEnhancer.close();
            } catch (CameraEnhancerException e) {
                e.printStackTrace();
            }
        }
        super.handleOnPause();
    }

    @Override
    protected void handleOnResume() {
        if (mCameraEnhancer!=null){
            try {
                if (previousCameraStatus == EnumCameraState.OPENED) {
                    mCameraEnhancer.open();
                }
            } catch (CameraEnhancerException e) {
                e.printStackTrace();
            }
        }
        super.handleOnResume();
    }
}
